fggs/indices.py

```diff
@@ -0,0 +1,1293 @@
+
+def broadcast(vembedss: Sequence[Sequence[Embedding]]) \
+    -> Tuple[Tuple[List[Embedding], List[EmbeddingVar]], ...]:
+    """Apply broadcast semantics to the given vembed sequences
+       to make them equal in length (maximum ndim) and size,
+       creating fresh EmbeddingVars for each broadcast dimension.
+       We also return lists of the fresh EmbeddingVars created."""
+    rets : Tuple[Tuple[List[Embedding], List[EmbeddingVar]], ...] \
+         = tuple(([], []) for vembed in vembedss)
+    for es in zip_longest(*(reversed(vembeds) for vembeds in vembedss),
+                          fillvalue=ProductEmbedding(())):
+        ns = tuple(e.numel() for e in es)
+        Ns = frozenset(n for n in ns if n != 1)
+        if len(Ns) > 1: raise RuntimeError(f"Size mismatch {ns}")
+        N = tuple(Ns)[0] if Ns else 1
+        for ret, e, n in zip(rets, es, ns):
+            if n != N:
+                k = EmbeddingVar(N)
+                e = k if e == ProductEmbedding(()) else ProductEmbedding((e, k))
+                ret[1].insert(0, k)
+            ret[0].insert(0, e)
+    return rets
+
+@dataclass
+class EmbeddedTensor:
+    physical: Tensor
+    pembeds: Sequence[EmbeddingVar] = cast(Sequence[EmbeddingVar], None) #-/
+    vembeds: Sequence[Embedding]    = cast(Sequence[Embedding],    None) #-/
+    default: NumberType = 0
+
+    def __post_init__(self):
+        # Default to the trivial (dense) embedding, to convert from Tensor
+        if self.pembeds is None:
+            self.pembeds = tuple(EmbeddingVar(n) for n in self.physical.size())
+        else:
+            if __debug__:
+                if self.physical.size() != Size(k.numel() for k in self.pembeds):
+                    raise ValueError(f"EmbeddedTensor(tensor of {self.physical.size()}, pembeds of {Size(k.numel() for k in self.pembeds)}, ...)")
+        if self.vembeds is None:
+            self.vembeds = self.pembeds
+
+    @staticmethod
+    def from_int(x: int, semiring: Semiring) -> EmbeddedTensor:
+        return EmbeddedTensor(semiring.from_int(x), default=semiring.from_int(0).item())
+
+    @staticmethod
+    def eye(size: int, semiring: Semiring) -> EmbeddedTensor:
+        k = EmbeddingVar(size)
+        return EmbeddedTensor(semiring.from_int(1).expand(size), (k,), (k,k),
+                              semiring.from_int(0).item())
+
+    def depict(self, letterer: Callable[[EmbeddingVar], str]) -> str:
+        P = (k.depict(letterer) for k in self.pembeds)
+        V = (e.depict(letterer) for e in self.vembeds)
+        return f"[\\{' '.join(P)} -> {', '.join(V)} | default {self.default}]"
+
+    @property
+    def shape(self):
+        return self.size()
+
+    def numel(self) -> int:
+        return reduce(mul, (e.numel() for e in self.vembeds), 1)
+
+    def size(self) -> Size:
+        return Size(e.numel() for e in self.vembeds)
+
+    def dim(self) -> int:
+        return len(self.vembeds)
+
+    ndim = ndimension = property(dim)
```

Comment:
Is ndimension ever used?

Reply:
It's not. I just put it in because I saw it next to ndim in the pytorch documentation. But oh wait, I got it wrong, `ndimension` is a method like `dim`, not a property like `ndim`! I'll fix that.